### PR TITLE
In memory index driver

### PR DESCRIPTION
**Pull request by SpacemanPaul**
### Changes

Implements a non-persistent, local, in-memory index driver.  The index is maintained in local memory and is not backed by a database.

This driver used by setting the ``index_driver`` to ``memory`` in the configuration file.

The memory index driver may be useful:

1. for ODC use cases where there is no need for the index to be re-used beyond the current session;
2. for testing scenarios where no index persistence is required; or
3. as an example/template for developing other index drivers.

### Implementation

Datasets are stored in the index as Dataset models.  Dataset models returned by the API are cloned from the model stored in the index, and can be modified without affecting the version stored in the index. API method implementations are based on the default driver implementations where possible and are intended to be as compatible with them as possible.

The only incompatibility I am aware of is that the in-memory driver cannot search by fields that are not contained in the dataset and product metadata documents (e.g. `added`, `added_by`, `updated`, `dataset_type_id`, etc.)

### Motivation

Further validation of the index driver interface, and to gain a deeper understanding of the index and model APIs in preparation for a more detailed roadmap to addressing [EP-003 for ODCv2](https://github.com/opendatacube/datacube-core/wiki/ODC-EP-003---Replace-the-ODC-Index-and-Internal-Database-API). 

### Status

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/index/abstract.py
Comment (Kirill888) [inline datacube/index/abstract.py]: `(object)` is now discouraged.

Comment (SpacemanPaul) [reply datacube/index/abstract.py]: Thanks, missed this one when I relocated the definition.  

I missed some other stuff too judging by the test results ...  ah, a more significant typo in a test fixture.

File: datacube/index/abstract.py
Comment (Kirill888) [inline datacube/index/abstract.py:750]: that reminds me what did we decide about "source_filter", is it "ok to not support that functionality but we are keeping it for postgres backend"?

Comment (SpacemanPaul) [reply datacube/index/abstract.py:750]: I'm not sure we *decided* anything yet.

My next major task after this is writing up a detailed ODCv2 roadmap/plan for review and discussion by the SC and broader community, as to how I think we should move forward. For this PR, I'm slavishly copying the postgres driver behaviour.  I suspect there will be a LOT of "ok to not support that functionality but we are keeping it for postgres backend" items in that plan.

**Comment by alexgleith** [comment]
This looks good to me, but I don't understand the code properly to have confidence. I trust you and if you need approval, let me know and I'll read things more carefully and approve :-) 

It's fantastic that you're tackling this, by the way Paul. Thankyou!

**Comment by woodcockr** [comment]
I've given this a solid look over but I'm a noob to many part of this code and using this as an opportunity to begin getting my head around it in order to participate in the upcoming development. The code seems clear, tests in order, reasoning sound. If I had more familiarity with this section I'd approve.

File: datacube/index/abstract.py
Comment (SpacemanPaul) [reply datacube/index/abstract.py:750]: I'm now strongly leaning towards dropping support for source_filter queries in ODCv2.

File: docs/installation/extending-odc.rst
Comment (omad) [inline docs/installation/extending-odc.rst:247]: Typo: should be datacube.index.memory

File: datacube/utils/documents.py
Comment (omad) [inline datacube/utils/documents.py:316]: Can you please add a docstring. the tests make it look as though it's recursively checking whether element is contained within document.

It looks like the third case when document is not a dict, and element is a list or tuple isn't covered by the tests, I'm not sure what it means.

File: datacube/model/_base.py
Comment (omad) [inline datacube/model/_base.py:10]: Probably worth having a docstring mentioning this is inclusive at the start and exclusive at the end.

I think that's the right terminology?

**Comment by SpacemanPaul** [comment]
> I've given this a solid look over but I'm a noob to many part of this code and using this as an opportunity to begin getting my head around it in order to participate in the upcoming development. The code seems clear, tests in order, reasoning sound. If I had more familiarity with this section I'd approve.

Honestly, a lot of the weirdest-looking stuff is a direct copy and paste from the postgres driver.

File: datacube/index/memory/_datasets.py
Comment (omad) [inline datacube/index/memory/_datasets.py:35]: All these `Mapping` types should be `MutableMapping`.

`Mapping` doesn't include `__setitem__` used for assignment with `[]`.

See: https://stackoverflow.com/questions/52487663/python-type-hints-typing-mapping-vs-typing-dict

**Review by omad — APPROVED** [review]
Great work Paul!

There's nothing I can see that requires changing before merging, but I've left a few comments of things that could be tidied up after.

Some of them are things that extend beyond this PR.
- We have a `pre-commit` configuration to automatically format code and organise imports, but it's not being checked and a lot of code doesn't meet it's defined standards.
- There's a lot of type hints, but there's quite a few which aren't correct enough to pass `mypy` checks, or whatever is built into `PyCharm`. Eg, the one I mentioned specifically is `Mapping` vs `MutableMapping`, but there's more.

**Pull request merged**
